src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -1,81 +1,131 @@
         } else {
-          val hash = bestBlock.header.hash
-          blockchain.getChainWeightByHash(hash) match {
+          blockchain.getChainWeightByHash(bestBlock.header.hash) match {
             case Some(weight) =>
-              val importResult = if (isPossibleNewBestBlock(block.header, bestBlock.header)) {
-                importToTop(block, bestBlock, weight)
-              } else {
-                reorganise(block, bestBlock, weight)
+              doBlockPreValidation(block).flatMap {
+                case Left(error) => Task.now(BlockImportFailed(error.reason.toString))
+                case Right(_)    => handleBlockImport(block, bestBlock, weight)
               }
-              importResult.foreach(measureBlockMetrics)
-              importResult
-            case None =>
-              log.error(
-                "Getting total difficulty for current best block with hash: {} failed",
-                bestBlock.header.hashAsHexString
-              )
-              Task.now(
-                BlockImportFailed(
-                  s"Couldn't get total difficulty for current best block with hash: ${bestBlock.header.hashAsHexString}"
-                )
-              )
+            case None => returnNoTotalDifficulty(bestBlock)
           }
         }
-      case None =>
-        log.error("Getting current best block failed")
-        Task.now(BlockImportFailed("Couldn't find the current best block"))
+      case None => returnNoBestBlock()
     }
 
+  private def handleBlockImport(block: Block, bestBlock: Block, weight: ChainWeight)(implicit
+      blockExecutionScheduler: Scheduler,
+      blockchainConfig: BlockchainConfig
+  ): Task[BlockImportResult] = {
+    val importResult = if (isPossibleNewBestBlock(block.header, bestBlock.header)) {
+      importToTop(block, bestBlock, weight)
+    } else {
+      reorganiseOrEnqueue(block, bestBlock, weight)
+    }
+    importResult.foreach(measureBlockMetrics)
+    importResult
+  }
+
+  private def doBlockPreValidation(block: Block)(implicit
+      blockchainConfig: BlockchainConfig
+  ): Task[Either[ValidationBeforeExecError, BlockExecutionSuccess]] =
+    Task
+      .evalOnce {
+        val validationResult = blockValidation.validateBlockBeforeExecution(block)
+        validationResult.left.foreach { error =>
+          log.error("Error while validating block with hash {} before execution: {}", block.hash, error.reason)
+        }
+        validationResult
+      }
+      .executeOn(validationScheduler)
+
+  private def returnNoTotalDifficulty(bestBlock: Block): Task[BlockImportFailed] = {
+    log.error(
+      "Getting total difficulty for current best block with hash: {} failed",
+      bestBlock.header.hashAsHexString
+    )
+    Task.now(
+      BlockImportFailed(
+        s"Couldn't get total difficulty for current best block with hash: ${bestBlock.header.hashAsHexString}"
+      )
+    )
+  }
+
+  private def returnNoBestBlock(): Task[BlockImportFailed] = {
```

Comment:
Why is this split out into a method? Doesn't seem to do much.

(If anything should be split out: `BlockImportFailed("Couldn't find the current best block")` could be a `val`.)

Perhaps we can reuse these error descriptions as log messages as well?

Reply:
It was mainly to make the main method `evaluateBranchBlock` more readable, by moving the detail of logging the error and creating the actual error description

Reply:
So I would argue it is less readable now since those methods actually do so little.